src/Dfc.CourseDirectory.WebV2/Behaviors/RequireUserCanAccessCourseBehavior.cs

```diff
@@ -0,0 +1,48 @@
+﻿using System.Threading;
+using System.Threading.Tasks;
+using Dfc.CourseDirectory.Core;
+using Dfc.CourseDirectory.WebV2.Security;
+using MediatR;
+
+namespace Dfc.CourseDirectory.WebV2.Behaviors
+{
+    public class RequireUserCanAccessCourseBehavior<TRequest, TResponse>
+        : IPipelineBehavior<TRequest, TResponse>
+    {
+        private readonly IRequireUserCanAccessCourse<TRequest> _descriptor;
+        private readonly ICurrentUserProvider _currentUserProvider;
+        private readonly IProviderOwnershipCache _providerOwnershipCache;
+
+        public RequireUserCanAccessCourseBehavior(
+            IRequireUserCanAccessCourse<TRequest> descriptor,
+            ICurrentUserProvider currentUserProvider,
+            IProviderOwnershipCache providerOwnershipCache)
+        {
+            _descriptor = descriptor;
+            _currentUserProvider = currentUserProvider;
+            _providerOwnershipCache = providerOwnershipCache;
+        }
+
+        public async Task<TResponse> Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate<TResponse> next)
+        {
+            var courseId = _descriptor.GetCourseId(request);
+            var providerId = await _providerOwnershipCache.GetProviderForCourse(courseId);
+
+            if (providerId == null)
+            {
+                throw new ResourceDoesNotExistException(ResourceType.Course, courseId);
+            }
+
+            var currentUser = _currentUserProvider.GetCurrentUser();
+
+            // If user is not an admin then check they belong to the same provider as the requested course
+            if ((currentUser.Role == RoleNames.ProviderUser || currentUser.Role == RoleNames.ProviderSuperUser) &&
```

Comment:
As discussed, an explicit check for `RoleNames.Developer/Helpdesk` would be preferable here.

Reply:
Potential security bug if `currenUser.Role` does not exist or is not `ProviderUser` or `ProviderSuperUser`.

Reply:
Other parts of the stack guarantee that the user is in exactly one of the `ProviderUser`, `ProviderSuperUser`, `Helpdesk` or `Developer` roles but I've added explicit checks